src/sodaspark/scan.py

```diff
@@ -297,3 +301,136 @@ def execute(
+        The measurements as data frame.
+    """
+    schema_group_values = T.StructType(
+        [
+            T.StructField(
+                "group", T.MapType(T.StringType(), T.StringType()), True
+            ),
+            T.StructField("value", T.StringType(), True),
+        ]
+    )
+    schema = T.StructType(
+        [
+            T.StructField("metric", T.StringType(), True),
+            T.StructField("columnName", T.StringType(), True),
+            T.StructField("value", T.StringType(), True),
+            T.StructField(
+                "groupValues", T.ArrayType(schema_group_values, True), True
+            ),
+        ]
+    )
+    spark_session = SparkSession.builder.getOrCreate()
+    out = spark_session.createDataFrame(
+        [measurement.to_dict() for measurement in measurements],
+        schema=schema,
+    )
+    return out
+
+
+def testresults_to_data_frame(testresults: list[TestResult]) -> DataFrame:
+    """
+    Convert TestResults to a data frame.
+    Parameters
+    ----------
+    testresults: List[TestResult]
+        The testresults.
+    Returns
+    -------
+    out : DataFrame
+        The testresults as data frame.
+    """
+
+    schema_test = T.StructType(
+        [
+            T.StructField("column", T.StringType(), True),
+            T.StructField("expression", T.StringType(), True),
+            T.StructField("id", T.StringType(), True),
+            T.StructField("metrics", T.ArrayType(T.StringType(), True), True),
+            T.StructField("title", T.StringType(), True),
+        ]
+    )
+
+    schema = T.StructType(
+        [
+            T.StructField("test", schema_test, True),
+            T.StructField("passed", T.BooleanType(), True),
+            T.StructField("skipped", T.BooleanType(), True),
+            T.StructField(
+                "values", T.MapType(T.StringType(), T.StringType()), True
+            ),
+            T.StructField("error", T.StringType(), True),
+            T.StructField(
+                "group_values", T.MapType(T.StringType(), T.StringType()), True
+            ),
+        ]
+    )
+    spark_session = SparkSession.builder.getOrCreate()
+    out = spark_session.createDataFrame(
+        testresults,
+        schema=schema,
+    )
+    return out
+
+
+def scanerror_to_data_frame(scanerrors: list[ScanError]) -> DataFrame:
+    """
+    Convert ScanError to a data frame.
+    Parameters
+    ----------
+    scanerror: List[ScanError]
+        The scanerrors.
+    Returns
+    -------
+    out : DataFrame
+        The scanerrors as data frame.
+    """
+    schema = T.StructType(
+        [
+            T.StructField("type", T.StringType(), True),
+            T.StructField("message", T.StringType(), True),
+            T.StructField("exception", T.StringType(), True),
+        ]
+    )
+    spark_session = SparkSession.builder.getOrCreate()
+    out = spark_session.createDataFrame(
+        [scanerror.to_dict() for scanerror in scanerrors], schema=schema
```

Comment:
Also remove the `to_dict` here

Reply:
I have resolved this by updating the schema. The `type` is not a field but instead a method. If we want to expose we would need to leverage `to_dict()`. I think its better to keep it as close to `ScanError` as possible.